python/ray/data/impl/plan.py

```diff
@@ -30,6 +30,7 @@ def __init__(self, in_blocks: BlockList, stats: DatasetStats, dataset_uuid=None)
         self._out_blocks = None
         self._in_stats = stats
         self._out_stats = None
+        self._out_stage = None
```

Comment:
```suggestion
        self._num_stages_executed = 0
```

Reply:
Actually reading through this again, I think it would be cleaner if we:
1. Added an additional `parent_plan` argument to ExecutionPlan.
2. If in_blocks is cleared, the parent plan can be executed to generate the in_blocks again.

This would remove all the scary off-by-one errors related to managing this counter.

Reply:
I thought about doing that! But I was concerned about unnecessary holding onto extra block references.

I'll try this out!

Reply:
Hmm that's a good point. Come to think of it, we may be holding unnecessary block references in the form of _in_blocks already.

What if we instead did this:
1. Add a new field called _base_tasks, which is just a copy of the read tasks from the base LazyBlockList if it exists.
2. Add a new field called _prev_stages, which is the list of stages needed to transform _base_tasks to _in_blocks.

That way, we can clear _in_blocks and know how to re-generate it from _base_tasks + _prev_stages.

Reply:
>Come to think of it, we may be holding unnecessary block references in the form of _in_blocks already.

I actually brought that up in a [past PR review](https://github.com/ray-project/ray/pull/22233#discussion_r803104432), was supposed to be fixed in the fusion PR but looks like it was missed.

>That way, we can clear _in_blocks and know how to re-generate it from _base_tasks + _prev_stages.

Yep that's kind of similar to what I have right now conceptually, where we have a block snapshot at some point in the stage lineage and we proceed from the snapshot if it exists or derive from the input blocks if it doesn't. In your case, we would just allow the input blocks to be either a set of read tasks or a block list.

I'll try out something like this that eliminates the snapshot index tracking.

Reply:
@ericl I started making these changes but the ideal solution involves refactoring `LazyBlockList` and its use, along with the execution plan, which might balloon this PR a bit. In an effort to limit the scope of this PR and to unblock the ML team, I'd like to go ahead with the current manual snapshot tracking; I cleaned up the snapshot tracking and left comments explaining any subtle.

We can refactor `LazyBlockList` and `ExecutionPlan` in a follow-up PR to remove the snapshot index bookkeeping and make the input block story nicer.